### PR TITLE
Feat/xcm

### PR DESCRIPTION
**Pull request by Kevingislason**
## Description

XCM prototype

---

Before we can merge this PR, please make sure that all the following items have been checked off. If any of the checklist items are not applicable, please leave them but write a little note why.

- [ ] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work
- [ ] Re-reviewed files changed in the Github PR explorer

## If PR to `staging`
- [ ] Updated relevant documentation in the code
- [ ] Test code changes manually, and describe your procedure in a comment on this PR
- [ ] Mark this PR with the correct milestone

## If PR to `main`
- [ ] Update the version number in `package.json` and `src/config/common.json`
- [ ] Update the minimum required signer version number in `src/config/common.json` if applicable
- [ ] Test following the procedure in `docs/manual_test.md`
- [ ] Update `CHANGELOG.md`
- [ ] Make sure that all PR's to `manta-signer` and `sdk` on which this PR depends are merged

### CODE REVIEW

File: src/components/Balance/index.tsx [hunk 1]
```diff
@@ -16,8 +16,8 @@ const Balance: React.FC<IBalanceProps> = ({
 }) => {
   return (
     <div id="balanceText" className={className}>
-      Balance:&nbsp;<strong>{balance}</strong>
```
Comment (Kevingislason) [inline src/components/Balance/index.tsx:19]: Rename 'BalanceDisplay'

File: src/components/Navbar/SignerConnectionStatusLabel.tsx [hunk 2]
```diff
@@ -63,7 +63,7 @@ const SignerConnectionStatusLabel = () => {
   const config = useConfig();
   const { signerVersion } = usePrivateWallet();
   return (
-    <div className="flex text-center items-center text-green-500 pr-6">
```
Comment (Kevingislason) [inline src/components/Navbar/SignerConnectionStatusLabel.tsx:66]: Delete

File: src/config/types.ts [hunk 3]
```diff
@@ -34,7 +34,7 @@ async function main () {
   const recordOpt = await api.query.voting.voteRecords(123);
```
Comment (Kevingislason) [inline src/config/types.ts:34]: does this file do anything?

File: src/contexts/privateWalletContext.tsx [hunk 4]
```diff
@@ -115,10 +115,12 @@ export const PrivateWalletContextProvider = (props) => {
   }, [api, externalAccountSigner, signerIsConnected, signerVersion]);
 
   const fetchSignerVersion = async () => {
+    console.log('fetchSignerVersion', fetchSignerVersion)
```
Comment (Kevingislason) [inline src/contexts/privateWalletContext.tsx:118]: delete

File: src/contexts/privateWalletContext.tsx [hunk 5]
```diff
@@ -115,10 +115,12 @@ export const PrivateWalletContextProvider = (props) => {
...
     try {
       const res = await axios.get(`${config.SIGNER_URL}version`, {
         timeout: 1500
       });
+      console.log('res', res);
```
Comment (Kevingislason) [inline src/contexts/privateWalletContext.tsx:123]: delete

File: src/eth/EthXCM.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  const accountId = addressToAccountId(address);
+  const destination = getXtokensPrecompileLocation(Chain.Calamari(config).parachainId, accountId);
+  const weight = CALAMARI_DESTINATION_WEIGHT;
+
+  try {
+    console.log('params', contract, amount, destination, weight);
+    console.log(
+      'estimate',
+      contract.estimateGas.transfer(ERC_PRECOMPILE_ADDRESS, amount, destination, weight)
+    )
```
Comment (Kevingislason) [inline src/eth/EthXCM.ts:70]: delete logs

File: src/pages/BridgePage/BridgeAssetSelect.tsx [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    setSelectedAssetType,
+  } = useBridgeData();
+  const {
+    txIsOverMinAmount,
+    userHasSufficientFunds,
+  } = useBridgeTx();
+
+  const balanceText = senderAssetCurrentBalance
+    ? `${senderAssetCurrentBalance.toString()} ${senderAssetType.ticker}`
+    : null;
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeAssetSelect.tsx:24]: '--'

File: src/pages/BridgePage/BridgeAssetSelect.tsx [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+  let errorText = null;
+  if (txIsOverMinAmount() === false) {
+    errorText = 'Tx amount too low'
+  } else if (userHasSufficientFunds() === false) {
+    errorText = 'Tx amount is too high'
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeAssetSelect.tsx:30]: This error text could be better

File: src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+    const getOriginFee = (inputConfig) => {
+      return new Balance(
+        originChain.nativeAsset,
+        new BN(inputConfig.estimateFee)
+      );
+    }
+    const getMaxInput = (inputConfig) => {
+      return Balance.fromBaseUnits(
+        senderAssetType,
+        Decimal.max(new Decimal(inputConfig.maxInput.toString()), new Decimal(0))
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx:106]: Should fix this in polkawallet bridge PR

File: src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+      const amount = senderAssetTargetBalance
+        ? senderAssetTargetBalance.valueBaseUnits.toString()
+        : "0";
+
+      let address = destinationAddress;
+      // Can't estimate fees for eth addresses like on Moonriver; use any substrate address instead
+      // if (destinationAddress === ethAddress) {
+      //   const ARBITRARY_ADDRESS = '5HDoTPBGGxfnkg6DNacyvCz6FzENJ2bgWkas239VfY9CGq72';
+      //   address = ARBITRARY_ADDRESS;
+      // };
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx:136]: delete

File: src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+  // These populate public address select dropdowns in the ui
+  useEffect(() => {
+    const initPublicAccountOptions = () => {
+      dispatch({
+        type: BRIDGE_ACTIONS.SET_SENDER_SUBSTRATE_ACCOUNT_OPTIONS,
+        senderSubstrateAccountOptions: externalAccountOptions
+      });
+    };
+    initPublicAccountOptions();
+  }, [externalAccountOptions]);
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx:178]: possibly not needed?

File: src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+
+  // Synchronizes the user's current 'active' public account in local state
+  // to match its upstream source of truth in `externalAccountContext`
+  // The active `senderSubstrateAccount` sends and covers fees for all
+  // substrate-based payments
+  useEffect(() => {
+    const syncSenderSubstrateAccountToExternalAccount = () => {
+      dispatch({
+          type: BRIDGE_ACTIONS.SET_SENDER_SUBSTRATE_ACCOUNT,
+          senderSubstrateAccount: externalAccount
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx:192]: not totally clear we need `senderSubstrateAccount` at all?

File: src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+  // to match the user's public account as set in the send form
+  useEffect(() => {
+    const syncExternalAccountToSubstrateAccount = () => {
+      senderSubstrateAccount && changeExternalAccount(senderSubstrateAccount);
+    };
+    syncExternalAccountToSubstrateAccount();
+  }, [
+    senderAssetType,
+    externalAccountOptions
+  ]);
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx:208]: not clear this is needed

File: src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+
+  /**
+   *
+   * Mutations exposed through UI
+   */
+
+  // Sets the sender's public account, exposed in the `To Public` and `Public transfer` form;
+  // State is set upstream in `externalAccountContext`, and propagates downstream here
+  // (see `syncPublicAccountToExternalAccount` above)
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx:217]: comment incorrect

File: src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+  const setSenderSubstrateAccount = async (senderSubstrateAccount) => {
+    setLastAccessedExternalAccountAddress(senderSubstrateAccount.address);
+    await changeExternalAccount(senderSubstrateAccount);
+  };
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx:221]: possibly not needed

File: src/pages/BridgePage/BridgeContext/BridgeTxContext.tsx [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+    );
+  };
+
+  /**
+   *
+   * Transactions
+   */
+
+  // Handles the result of a transaction
+  const handleTxRes = async ({ status, events }) => {
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/BridgeTxContext.tsx:93]: check that this logs the failed extrinsic error correctly

File: src/pages/BridgePage/BridgeContext/BridgeTxContext.tsx [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+    destinationChain,
+    originAddress,
+    destinationAddress,
+    maxInput,
+    minInput
+  } = useBridgeData();
+
+  /**
+   *
+   * Transaction validation
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/BridgeTxContext.tsx:39]: May also need to check that user has sufficient funds to pay tx fee?

File: src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+    senderAssetType,
+    senderAssetTargetBalance,
+    senderSubstrateAccount,
+    originChainOptions,
+    originChain,
+    destinationChain,
+    bridge,
+  } = state;
+
+  const originAddress = originChain.ethMetadata ? ethAddress : senderSubstrateAccount?.address;
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/BridgeDataContext.tsx:42]: `originChain.xcmAdapter.chain.type === 'ethereum'` is probably a better check

File: src/pages/BridgePage/BridgeContext/BridgeTxContext.tsx [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+      await tx.signAndSend(externalAccountSigner, handleTxRes);
+    } catch (error) {
+      console.error('Transaction failed', error);
+      setTxStatus(TxStatus.failed());
+    }
+  };
+
+  // Attempts to build and send a bridge transaction with an Eth-like origin chain
+  const sendEth = async () => {
+    const success = await transferMovrFromMoonriverToCalamari(
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/BridgeTxContext.tsx:154]: do a check first

File: src/pages/BridgePage/BridgeContext/bridgeActions.tsx [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+  SET_FEE_ESTIMATES: 'SET_FEE_ESTIMATES',
+  SET_SENDER_ASSET_TARGET_BALANCE: 'SET_SENDER_ASSET_TARGET_BALANCE',
+  SET_SENDER_NATIVE_TOKEN_PUBLIC_BALANCE: 'SET_SENDER_NATIVE_TOKEN_PUBLIC_BALANCE',
+  SET_SENDER_ETH_ACCOUNT: 'SET_SENDER_ETH_ACCOUNT',
+  SET_SENDER_SUBSTRATE_ACCOUNT: 'SET_SENDER_SUBSTRATE_ACCOUNT',
+  SET_SENDER_SUBSTRATE_ACCOUNT_OPTIONS: 'SET_SENDER_PUBLIC_ACCOUNT_OPTIONS',
+  SET_ORIGIN_CHAIN: 'SET_ORIGIN_CHAIN',
+  SET_DESTINATION_CHAIN: 'SET_DESTINATION_CHAIN',
+  SET_CHAIN_OPTIONS: 'SET_CHAIN_OPTIONS',
+  SET_BRIDGE: 'SET_BRIDGE'
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/bridgeActions.tsx:16]: check that all of these are actually used

File: src/pages/BridgePage/BridgeContext/bridgeReducer.tsx [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+
+    originChain: initOriginChain,
+    originChainOptions: initOriginChainOptions,
+    originFee: null,
+
+    destinationChain: initDestinationChain,
+    destinationChainOptions: initDestinationChainOptions,
+    destinationFee: null,
+  };
+}
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/bridgeReducer.tsx:73]: Move init logic to its own file

File: src/pages/BridgePage/BridgeContext/bridgeReducer.tsx [hunk 22]
```diff
@@ -0,0 +1,264 @@
...
+
+  case BRIDGE_ACTIONS.SET_DESTINATION_CHAIN:
+    return setDestinationChain(state, action);
+
+  case BRIDGE_ACTIONS.SET_CHAIN_OPTIONS:
+    return setChainOptions(state, action);
+
+  default:
+    throw new Error(`Unknown type: ${action.type}`);
+  }
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/bridgeReducer.tsx:114]: check all of these actions are being used

File: src/pages/BridgePage/BridgeContext/bridgeReducer.tsx [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+};
+
+const balanceUpdateIsStale = (stateAssetType, updateAssetType) => {
+  if (!updateAssetType) {
+    return false;
+  }
+  return stateAssetType?.assetId !== updateAssetType.assetId;
+};
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/bridgeReducer.tsx:122]: check if needed

File: src/pages/BridgePage/BridgeContext/bridgeReducer.tsx [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+
+const setSenderNativeTokenPublicBalance = (state, action) => {
+  return {
+    ...state,
+    senderNativeTokenPublicBalance: action.senderNativeTokenPublicBalance
+  };
+};
+
+const setFeeEstimates = (state, action) => {
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeContext/bridgeReducer.tsx:187]: rename

File: src/pages/BridgePage/BridgeFeeDisplay.tsx [hunk 25]
```diff
@@ -0,0 +1,25 @@
+// @ts-nocheck
+import React from 'react';
+import { useBridgeData } from './BridgeContext/BridgeDataContext';
+
+const BridgeFeeDisplay = () => {
+  const { originFee, destinationFee } = useBridgeData();
+
+  const originFeeText = originFee ? originFee.toFeeString() : '';
+  const destinationFeeText = destinationFee ? destinationFee.toFeeString() : '';
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeFeeDisplay.tsx:9]: '--'

File: src/components/AmountInput/BridgeBalanceInput.tsx [hunk 26]
```diff
@@ -0,0 +1,71 @@
+// @ts-nocheck
```
Comment (Kevingislason) [inline src/components/AmountInput/BridgeBalanceInput.tsx:1]: move into bridge page

File: src/components/AmountInput/SendBalanceInput.tsx [hunk 27]
```diff
@@ -0,0 +1,66 @@
+// @ts-nocheck
```
Comment (Kevingislason) [inline src/components/AmountInput/SendBalanceInput.tsx:1]: Move into send page

File: src/pages/BridgePage/BridgeForm.tsx [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+            chain={originChain}
+            chainOptions={originChainOptions}
+            setChain={setOriginChain}
+            chainFromOrTo="From"
+          />
+          <img
+            className="mx-auto pb-7"
+            src={Svgs.ArrowRightIcon}
+            alt="switch-icon"
+          />
```
Comment (Kevingislason) [inline src/pages/BridgePage/BridgeForm.tsx:36]: Implement

File: src/pages/BridgePage/ChainSelect.tsx [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+
+  return (
+    <div>
+      <Select
+        className={classNames(
+          'w-52 rounded-lg manta-bg-gray text-black dark:text-white',
+          { disabled: disabled }
+        )}
+        isSearchable={false}
+        chainFromOrTo={chainFromOrTo}
```
Comment (Kevingislason) [inline src/pages/BridgePage/ChainSelect.tsx:28]: document

File: src/pages/BridgePage/MetamaskAccountDisplay.tsx [hunk 30]
```diff
@@ -0,0 +1,14 @@
+// @ts-nocheck
```
Comment (Kevingislason) [inline src/pages/BridgePage/MetamaskAccountDisplay.tsx:1]: Delete

File: src/pages/SendPage/SendContext/index.tsx [hunk 31]
```diff
@@ -73,6 +71,20 @@ export const SendContextProvider = (props) => {
     }
   }, [privateAddress]);
 
+  useEffect(() => {
+    const initReceiver = (receiverAddress) => {
```
Comment (Kevingislason) [inline src/pages/SendPage/SendContext/index.tsx:75]: document

File: src/pages/SendPage/SendFromForm/FromAccountSelect.tsx [hunk 32]
```diff
@@ -0,0 +1,80 @@
+// @ts-nocheck
```
Comment (Kevingislason) [inline src/pages/SendPage/SendFromForm/FromAccountSelect.tsx:1]: delete

File: src/pages/SendPage/SendFromForm/SendAssetSelect.tsx [hunk 33]
```diff
@@ -1,13 +1,46 @@
...
+  } = useSend();
+
+  const maxSendableBalance = getMaxSendableBalance();
+
+  const balanceText =
+    isInitialSync && (isPrivateTransfer() || isToPublic())
+      ? 'Syncing to ledger'
+      : senderAssetCurrentBalance
+        ? `${senderAssetCurrentBalance.toString()} ${senderAssetType.ticker}`
+        : '';
```
Comment (Kevingislason) [inline src/pages/SendPage/SendFromForm/SendAssetSelect.tsx:28]: avoid nested ternery

File: src/pages/SendPage/SendToForm/ReceiverBalanceDisplay.tsx [hunk 34]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
```
Comment (Kevingislason) [inline src/pages/SendPage/SendToForm/ReceiverBalanceDisplay.tsx:1]: Ask designers if we still want this

File: src/pages/SendPage/SendToForm/SendToAddressInput.tsx [hunk 35]
```diff
@@ -0,0 +1,87 @@
...
+    } else {
+      setReceiver(null);
+      const addressIsPrivate = validatePrivateAddress(event.target.value);
+      if (addressIsPrivate) {
+        setErrorMessage('Provided address is private, not public');
+      } else {
+        setErrorMessage('Invalid address');
+      }
+    }
+  };
```
Comment (Kevingislason) [inline src/pages/SendPage/SendToForm/SendToAddressInput.tsx:57]: deduplicate

File: src/pages/SendPage/SendToForm/SendToAddressInput.tsx [hunk 36]
```diff
@@ -0,0 +1,87 @@
...
+          disabled={disabled}
+          placeholder={'address'}
+        />
+      </div>
+      <p
+        className={`text-xss text-red-500 ml-2 ${
+          errorMessage ? 'visible' : 'invisible'
+        }`}
+      >
+        {errorMessage ?? 'No Error'}
```
Comment (Kevingislason) [inline src/pages/SendPage/SendToForm/SendToAddressInput.tsx:81]: Get rid of this hack

File: src/types/AssetType.ts [hunk 37]
```diff
@@ -2,16 +2,43 @@
...
+  PHA: 13
+}
+
+const getAssetIds = (config) => {
+  return config.NETWORK_NAME === 'Calamari' ? CalamariAssetIds : DolphinAssetIds;
+}
+
 export default class AssetType {
   constructor(
+    config,
```
Comment (Kevingislason) [inline src/types/AssetType.ts:33]: seems not to be needed

File: src/types/AssetType.ts [hunk 38]
```diff
@@ -2,16 +2,43 @@
 import Svgs from 'resources/icons';
 import BN from 'bn.js';
 
+const DolphinAssetIds = {
```
Comment (Kevingislason) [inline src/types/AssetType.ts:5]: ZK > P

File: src/types/AssetType.ts [hunk 39]
```diff
@@ -185,4 +250,12 @@ export default class AssetType {
       return this.toPrivate();
     }
   }
+
+  canTransferXcm = (originChain, destinationChain) => {
+    const res = (
+      originChain.xcmAssets.find(asset => asset.name === this.name)
+      && destinationChain.xcmAssets.find(asset => asset.name === this.name)
+    );
+    return res;
```
Comment (Kevingislason) [inline src/types/AssetType.ts:259]: no need for variable assignment

File: src/types/Balance.ts [hunk 40]
```diff
@@ -59,6 +59,16 @@ export default class Balance {
         })} ${this.assetType.ticker}`;
   }
 
+  toFeeString() {
+    return `${this.valueBaseUnits()
+      .toDecimalPlaces(6, Decimal.ROUND_UP)
```
Comment (Kevingislason) [inline src/types/Balance.ts:64]: I think this duplicates another display function; parameterize

File: src/types/Chain.ts [hunk 41]
```diff
@@ -0,0 +1,190 @@
...
+        Chain.Karura(config),
+        Chain.Moonriver(config)
+      ];
+    } else {
+      return [];
+    }
+  }
+
+  async _initApi() {
+    console.log("this.socket", this.socket)
```
Comment (Kevingislason) [inline src/types/Chain.ts:162]: delete

**Review by Kevingislason — COMMENTED** [review]
- [x] Add error handling
- [ ] Fix lint errors
- [x] Check validations for bridge component
- [x] Check on multiple screen widths
- [ ] Add comments
- [x] Implement toggle origin / destination chain

File: src/pages/BridgePage/BridgeContext/bridgeReducer.tsx (hunk 21 above)
Comment (Kevingislason) [reply src/pages/BridgePage/BridgeContext/bridgeReducer.tsx:73]: nah bad idea

File: src/components/Accounts/AccountSelectDropdown.tsx [hunk 42]
```diff
@@ -5,38 +5,35 @@ import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
...
 import classNames from 'classnames';
-import { useSend } from '../SendContext';
 
-const PublicFromAccountSelect = () => {
-  const {
-    senderPublicAccount,
-    senderPublicAccountOptions,
-    setSenderPublicAccount
-  } = useSend();
+export const substrateAccountToReactSelectOption = (account) => {
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/components/Accounts/AccountSelectDropdown.tsx:9]: revisit after merging account select PR

File: src/components/Accounts/AccountSelectMenu.tsx [hunk 43]
```diff
@@ -0,0 +1,154 @@
...
+              )}
+              onClick={onClickAccountSelect}
+            >
+              <img className="w-6 h-6" src={PolkadotIcon} alt="Polkadot" />
+              {externalAccount.meta.name}
+            </div>
+            {showAccountList ? (
+              <div className="mt-3 bg-secondary rounded-3xl p-6 pr-2 absolute right-0 top-full z-50 border border-manta-gray">
+                <div className="max-h-96 overflow-y-scroll pr-4">
+                  {externalAccountOptions.map((account: any, index: number) => (
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/components/Accounts/AccountSelectMenu.tsx:71]: revisit after merging account select PR

File: src/components/Accounts/AccountSelectMenu.tsx [hunk 44]
```diff
@@ -0,0 +1,154 @@
...
+                          <div
+                            onClick={() => {
+                              changeExternalAccount(account);
+                              setShowAccountList(false);
+                            }}
+                          >
+                            {`${account.address.slice(
+                              0,
+                              4
+                            )}...${account.address.slice(-5)}`}
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/components/Accounts/AccountSelectMenu.tsx:95]: revisit after merging account select PR

File: src/components/Accounts/AccountSelectMenu.tsx [hunk 45]
```diff
@@ -0,0 +1,154 @@
...
+
+  const [showAccountList, setShowAccountList] = useState(false);
+  const [addressCopied, setAddressCopied] = useState(-1);
+  const disabled = txStatus?.isProcessing();
+
+  const copyToClipboard = (address: string, index: number) => {
+    navigator.clipboard.writeText(address);
+    setAddressCopied(index);
+    return false;
+  };
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/components/Accounts/AccountSelectMenu.tsx:34]: revisit after merging account select PR

File: src/components/Accounts/AccountSelectMenu.tsx [hunk 46]
```diff
@@ -0,0 +1,154 @@
...
+                            theme="polkadot"
+                          />
+                          {account.meta.name}
+                        </div>
+                        <div className="flex items-center gap-2 mt-1">
+                          <div
+                            onClick={() => {
+                              changeExternalAccount(account);
+                              setShowAccountList(false);
+                            }}
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/components/Accounts/AccountSelectMenu.tsx:90]: revisit after merging account select PR

File: src/components/Accounts/AccountSelectMenu.tsx [hunk 47]
```diff
@@ -0,0 +1,154 @@
...
+                            />
+                          )}
+                        </div>
+                      </div>
+                      <div
+                        className="border-2 font-semibold border-secondary text-link rounded-full py-1 px-2 cursor-pointer"
+                        onClick={() => {
+                          changeExternalAccount(account);
+                          setShowAccountList(false);
+                        }}
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/components/Accounts/AccountSelectMenu.tsx:126]: revisit after merging account select PR

File: src/components/Navbar/SignerConnectionStatusLabel.tsx (hunk 2 above)
Comment (Kevingislason) [reply src/components/Navbar/SignerConnectionStatusLabel.tsx:66]: revisit after private wallet PR

File: src/pages/BridgePage/BridgeContext/bridgeReducer.tsx (hunk 24 above)
Comment (Kevingislason) [reply src/pages/BridgePage/BridgeContext/bridgeReducer.tsx:187]: can't think of anything better tbh, this is not bad

File: src/pages/SendPage/SendFromForm/FromAccountSelect.tsx (hunk 32 above)
Comment (Kevingislason) [reply src/pages/SendPage/SendFromForm/FromAccountSelect.tsx:1]: revisit after merge private payment redesign PR

File: src/types/AssetType.ts [hunk 48]
```diff
@@ -2,6 +2,32 @@
...
+  KMA: 1,
+  KAR: 8,
+  AUSD: 9,
+  LKSM: 10,
+  MOVR: 11,
+  KSM: 12,
+  PHA: 13
+}
+
+const getAssetIds = (config) => {
```
Comment (AlexParshh) [inline src/types/AssetType.ts:27]: just a thought but might want to change the implementation here to support case where config.NETWORK_NAME === 'Manta'

File: src/pages/BridgePage/BridgeBalanceInput.tsx [hunk 49]
```diff
@@ -0,0 +1,71 @@
...
+          senderAssetType,
+          new Decimal(value)
+        );
+        setInputValue(value);
+        if (targetBalance.valueAtomicUnits.gt(new BN(0))) {
+          setSenderAssetTargetBalance(targetBalance);
+        } else {
+          setSenderAssetTargetBalance(null);
+        }
+      } catch (error) {}
```
Comment (AlexParshh) [inline src/pages/BridgePage/BridgeBalanceInput.tsx:41]: missing console.errror(error);

File: src/pages/BridgePage/BridgeContext/BridgeTxContext.tsx [hunk 50]
```diff
@@ -0,0 +1,199 @@
...
+
+  // Handles the result of a transaction
+  const handleTxRes = async ({ status, events }) => {
+    const api = originChain.api;
+    if (status.isInBlock) {
+      for (const event of events) {
+        if (api.events.system.ExtrinsicFailed.is(event.event)) {
+          const error = event.event.data[0];
+          if (error.isModule) {
+            console.log('phew')
```
Comment (AlexParshh) [inline src/pages/BridgePage/BridgeContext/BridgeTxContext.tsx:111]: not sure if this was used for testing

File: src/pages/BridgePage/SendButton.tsx [hunk 51]
```diff
@@ -0,0 +1,38 @@
...
+  const { send } = useBridgeTx();
+
+  const onClick = () => {
+    send();
+  };
+
+  return (
+
+    <div >
+      {txStatus?.isProcessing() ? (
```
Comment (AlexParshh) [inline src/pages/BridgePage/SendButton.tsx:20]: might be cleaner to replace `txStatus?.isProcessing()` with `disabled`

File: src/pages/SendPage/SendContext/sendReducer.ts [hunk 52]
```diff
@@ -88,9 +107,15 @@ const balanceUpdateIsStale = (stateAssetType, updateAssetType) => {
 
 
 const toggleSenderIsPrivate = (state) => {
+  console.log('3')
```
Comment (AlexParshh) [inline src/pages/SendPage/SendContext/sendReducer.ts:110]: might want to remove

File: src/pages/SendPage/SendContext/sendReducer.ts [hunk 53]
```diff
@@ -105,20 +130,31 @@ const toggleSenderIsPrivate = (state) => {
 };
 
 const toggleReceiverIsPrivate = (state) => {
+  console.log('2')
```
Comment (AlexParshh) [inline src/pages/SendPage/SendContext/sendReducer.ts:133]: same here

File: src/config/types.ts [hunk 54]
```diff
@@ -2,6 +2,7 @@
 // We need to import the augmented definitions "somewhere" in our project, however since we have
 // it in tsconfig as an override and the api/types has imports, it is not strictly required here.
 // Because of the tsconfig override, we could import from '@polkadot/{api, types}/augment'
+console.log('runs?')
```
[the remark this answers is not in the document]
Comment (Kevingislason) [reply src/config/types.ts:5]: was testing if this file even runs, which it doesn't, deleted

**Comment by Kevingislason** [comment]
Very annoyingly can't figure out signing this old commit
see https://github.com/Manta-Network/manta-front-end/pull/318

**Pull request closed without merging**